Automattic/wp-calypso
# Prevent duplicate sites in signup

#### Changes proposed in this Pull Request

* Detect if signup flow was loaded via back button, and in that case do not create a new site when the signup flow is complete.
* Check pau2Xa-1Io-p2#comment-6759 for an explanation on how this implementation works
* There's an edge case in which a user could end up seeing the launch flow UI version when going to wordpress.com/start to add a new site. But it's going to be an extremely rare case, and a few things have to align, so we can safely not bother about this issue:
    - Firstly, when the user is on checkout page from signup flow, they should navigate to a different part of Calypso on the _same tab_ without using the close button on checkout header, or empty cart action.
    - Next, on the _same tab_, they should navigate to wordpress.com/start by directly typing the URL in the address bar, or clicking a link that opens the page(on the same tab).
    - All of this should happen in < 1 day since a new site was created (signup destination cookie will expire in 1 day).
Important to note that this issue will not be a permanent problem - the user just needs to open wordpress.com/start in a new tab to see the regular signup flow again. Again, this is a super edge case and not something to worry about IMO.

#### Testing instructions



**Scenario 1** (Verify that going back from checkout page to signup, you can select a new domain/plan and complete purchase without creating a duplicate site).

1. Starting logged out of wpcom, go through the signup at wordpress.com/start and select a paid plan.
2. On the checkout page, hit browser back. Verify that you see the launch flow UI version:

<img width="1245" alt="Screenshot 2020-08-27 at 5 38 18 PM" src="https://user-images.githubusercontent.com/1269602/91440367-1eeaef00-e88c-11ea-97bd-1a48f85ab060.png">

3. Test the following two scenarios:
    (A) Selecting a free plan should take you to customer home and duplicate site should not be created. 
    (B) Select a custom domain and paid plan and complete payment. Verify that the new domain and plan got added to your site and that duplicate sites were not created.

**Scenario 2** (When on signup via browser back from checkout, page reload should not affect the UI)

1.  Starting logged out of wpcom, go through the signup at wordpress.com/start and select a paid plan.
2.  On the checkout page, hit browser back. Verify that you see the launch flow UI version (screen shot in Scenario 1)
3. Reload the page and verify that you still see the launch flow UI version.

**Scenario 3** (Add new site should work, and open the regular signup flow)

1.  Starting logged out of wpcom, go through the signup at wordpress.com/start and select a paid plan.
2.  On the checkout page, click the close button in the masterbar or empty all cart items.
3. Click "Add new site" from Calypso's sidebar, and verify that the domain step shows the regular signup flow UI(and not the launch flow UI)

**Scenario 4**
1. Try browser back from checkout in other flows besides `onboarding`, like in /start/premium, or the plan-domain swap flow*(for this you need to be assigned to the test variant, check https://github.com/Automattic/wp-calypso/pull/43675#pullrequestreview-440991033 for details on how to enter the variant).
2. In all these flows, you should be able to add purchase to the existing site without creating duplicates sites.

**Scenario 5** (When signup destination cookie is not cleared, closing tab and reopening a new tab should kill `sessionStorage` so /start should open "Add new site" flow instead of launch flow UI version)

1.  Starting logged out of wpcom, go through the signup at wordpress.com/start and select a paid plan.
2. On the checkout page, close tab and in a new tab, open wordpress.com/start
3. You should be taken to the "Add new site" flow to create a new site, with the regular signup UI.

**Scenario 6**

Verify that https://github.com/Automattic/wp-calypso/issues/45862 is fixed.
